packages/react-dom/src/__tests__/ReactDOMFizzServer-test.js

```diff
@@ -1478,4 +1478,85 @@ describe('ReactDOMFizzServer', () => {
     // We should've been able to display the content without waiting for the rest of the fallback.
     expect(getVisibleChildren(container)).toEqual(<div>Hello</div>);
   });
+
+  // @gate experimental
+  it('should respect unstable_avoidThisFallback', async () => {
+    let resolveInnerPromise;
+    const innerPromise = new Promise((res, rej) => {
+      resolveInnerPromise = res;
+    });
+    let innerPromisesResolved = false;
+
+    const InnerComponent = ({isClient}) => {
+      if (isClient) {
+        // Resuspend after re-rendering on client to check that fallback shows on client
+        throw new Promise(() => {});
+      }
+      if (!innerPromisesResolved) {
+        throw innerPromise;
+      }
+      return <Text text="inner component resolved" />;
+    };
+
+    function App({isClient}) {
+      return (
+        <div>
+          <Text text="Non Suspense Content" />
+          <Suspense
+            fallback={
+              <span>
+                <Text text="Loading Outer" />
+              </span>
+            }
+            unstable_avoidThisFallback={true}>
+            <span>
+              <InnerComponent isClient={isClient} />
+            </span>
+          </Suspense>
+        </div>
+      );
+    }
+
+    await act(async () => {
+      const {startWriting} = ReactDOMFizzServer.pipeToNodeWritable(
+        <App isClient={false} />,
+        writable,
+      );
+      startWriting();
+    });
+
+    expect(getVisibleChildren(container)).toEqual(
+      <div>Non Suspense Content</div>,
+    );
+
+    await act(async () => {
+      innerPromisesResolved = true;
+      resolveInnerPromise();
+    });
+
+    expect(getVisibleChildren(container)).toEqual(
+      <div>
+        Non Suspense Content<span>inner component resolved</span>
+      </div>,
+    );
+
+    await act(async () => {
+      const root = ReactDOM.createRoot(container, {hydrate: true});
+      // TODO: Figure out why it takes two renders to get suspense fallback to show
```

Comment:
I suspect its due to the existing `unstable_avoidThisFallback` on the client. I need to figure out how to bypass it in the case where the fallback was already avoided in SSR

Reply:
We should use the new hydrateRoot api instead of createRoot. That makes it a bit clearer that the first render is actually the hydration and the second render is an update.

The second update applies an update to the child tree, which doesn’t actually change anything, but React doesn’t know it won’t change anything because it can’t execute it to figure out. So instead it shows the fallback until it can. For the first render, the hydration pass, nothing could’ve changed so React doesn’t have to show the fallback. Whatever came from the server is still relevant.

To make this a bit clearer you can change the second render pass to actually render something different and assert that it’s not showing stale content.